packages/common/src/converter/protobuf-payload-converters.ts

```diff
@@ -121,17 +121,57 @@ export class ProtobufJsonPayloadConverter extends ProtobufPayloadConverter {
       return undefined;
     }
 
-    const jsonValue = protoJsonSerializer.toProto3JSON(value);
-
-    return this.constructPayload({
-      messageTypeName: getNamespacedTypeName(value.$type),
-      message: encode(JSON.stringify(jsonValue)),
-    });
+    const hasBufferChanged = setBufferInGlobal();
+    try {
+      const jsonValue = protoJsonSerializer.toProto3JSON(value);
+
+      return this.constructPayload({
+        messageTypeName: getNamespacedTypeName(value.$type),
+        message: encode(JSON.stringify(jsonValue)),
+      });
+    } finally {
+      resetBufferInGlobal(hasBufferChanged);
+    }
   }
 
   public fromPayload<T>(content: Payload): T {
-    const { messageType, data } = this.validatePayload(content);
-    return protoJsonSerializer.fromProto3JSON(messageType, JSON.parse(decode(data))) as unknown as T;
+    const hasBufferChanged = setBufferInGlobal();
+    try {
+      const { messageType, data } = this.validatePayload(content);
+      const res = protoJsonSerializer.fromProto3JSON(messageType, JSON.parse(decode(data))) as unknown as T;
+      replaceBuffers(res);
+      return res;
+    } finally {
+      resetBufferInGlobal(hasBufferChanged);
+    }
+  }
+}
+
+function replaceBuffers<X>(obj: X) {
```

Comment:
It's unfortunate that we have to do this. And BTW, we will need to iterate over arrays too, not just objects.

I looked at the source code for the proto json library and it seems like they only use the Buffer class for base64 conversion.
Could you please submit an issue to request that https://github.com/googleapis/proto3-json-serializer-nodejs be made browser compatible and not depend on `Buffer`? We could suggest that they use https://www.npmjs.com/package/base64-js for base64 conversion, which is used by the Buffer polyfill package.
This would be the ideal solution IMHO, we could of course also do the work and submit a PR to the proto3-json-serializer library.

In the meantime, let's keep this unfortunate work around.

Reply:
This should already iterate over arrays. Their type is 'object', entries() gives you the array index, and you can index with a string an array... But performance for large arrays will be horrible, so let me fix that...

As we discussed, the protojson repo looks like it is in maintenance mode, so I have not much hope, but it is worth a try!

I think that long term we will have to support a set of nodejs-agnostic polyfills for user code, and rely less on third-party dependencies for the core, e.g., fork proto3-json...

Reply:
I just filed an issue googleapis/proto3-json-serializer-nodejs#74 about direct browser support, will see...